### PR TITLE
Advanced Building Blocks - Enumerable Methods

### PR DESCRIPTION
**Pull request by Samitti**
I have finished creating a script file to house my enumerable methods. Kindly I ask for the review of my project.

### CODE REVIEW

File: my_enum.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
+module Enumerable
+  # my_each
+  def my_each
+    each do |item|
```
Comment (mosesogwo) [inline my_enum.rb]: You shouldn't use the original each in this method. You can use another kind of loop.
image

File: my_enum.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      yield(item)
+    end
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    i = 0
+    each do |item|
```
Comment (mosesogwo) [inline my_enum.rb]: Use your own my_each method instead of the ruby default each method.

File: my_enum.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    result = []
+    my_each do |item|
+      result << item if yield(item)
+    end
+    result
+  end
+
+  # my_all
+  def my_all?(args = nil)
+    if !args.nil? # obj.kind_of?(mod)
```
Comment (mosesogwo) [inline my_enum.rb]: Remove commented code

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @Samitti:clap:, good work on this:+1:. Let's make the following adjustments:**

Ensure that the following scenarios are covered in your methods 
- **my_each** should return an enumerator when no block is given
- **my_each_with_index** should return an enumerator when no block is given
- **my_select** should return an enumerator when no block is given
- **my_map** should return an enumerator when no block is given
-- To take care of the above, include a condition in each method that check if no block is given. See this [documentation](https://ruby-doc.org/core-2.6.5/Enumerable.html) for their behaviors when no block is given.

- **my_map** should return an array of the result of the evaluation of the block or proc on each element in the given enum. For example `[1, 3, 4, 5].my_map {|i| i < 2 }` should return `[true, false, false, false]` instead of `[1]`. Your current my_map method works like the select method which is not correct.

- For **my_inject**, you should take care of when a symbol is passed as an argument. 
`(5..10).my_inject(:+)`  should return `45` instead of `:+` that it currently returns. 


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Comment by Samitti** [comment]
I have fixed the issue raised above and now I hope everything is fine. Kindly ask for second review.

File: my_enum.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    result = []
+    my_each do |item|
+      if block_given?
+        result << yield(item) ? true : false
+      else
+        result << proc.call(item) ? true : false
+      end
+    end
+    result
+  end
```
Comment (stratospherique) [inline my_enum.rb]: * [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map #error
   # NoMethodError:
       undefined method `call' for nil:NilClass
     # ./my_enum.rb:99:in `block in my_map'
     # ./my_enum.rb:9:in `my_each'
     # ./my_enum.rb:95:in `my_map'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive Progress @Samitti :clap: :clap: However, some methods need further refactoring in order to get approval for your project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :pray: 

### That's it :relieved: When finished, kindly submit another code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by Samitti** [comment]
I have fixed the issue raised above and now I hope everything is fine. Kindly ask for a review.

**Review by pelzolga123 — APPROVED** [review]
Approved.

**Comment by pelzolga123** [comment]
## Great work, Congratulation :+1: Project Approved.
You did all changes required previously. Go ahead and merge to development.
_Goodluck in your next project_ *Happy coding*

**Pull request merged**